mne/source_estimate.py

```diff
@@ -2454,260 +2249,52 @@ def morph_data(subject_from, subject_to, stc_from, grade=5, smooth=None,
     if not isinstance(stc_from, _BaseSurfaceSourceEstimate):
         raise ValueError('Morphing is only possible with surface or vector '
                          'source estimates')
-
-    logger.info('Morphing data...')
-    subjects_dir = get_subjects_dir(subjects_dir, raise_error=True)
-    nearest = grade_to_vertices(subject_to, grade, subjects_dir, n_jobs)
-    tris = _get_subject_sphere_tris(subject_from, subjects_dir)
-    maps = read_morph_map(subject_from, subject_to, subjects_dir)
-
-    # morph the data
-    data = [stc_from.lh_data, stc_from.rh_data]
-    data_morphed = [None, None]
-
-    n_chunks = ceil(stc_from.data.shape[1] / float(buffer_size))
-
-    parallel, my_morph_buffer, _ = parallel_func(_morph_buffer, n_jobs)
-
-    for hemi in [0, 1]:
-        e = mesh_edges(tris[hemi])
-        e.data[e.data == 2] = 1
-        n_vertices = e.shape[0]
-        e = e + sparse.eye(n_vertices, n_vertices)
-        idx_use = stc_from.vertices[hemi]
-        if len(idx_use) == 0:
-            continue
-        data_morphed[hemi] = np.concatenate(
-            parallel(my_morph_buffer(data_buffer, idx_use, e, smooth,
-                                     n_vertices, nearest[hemi], maps[hemi],
-                                     warn=warn)
-                     for data_buffer
-                     in np.array_split(data[hemi], n_chunks, axis=1)), axis=1)
-
-    vertices = [nearest[0], nearest[1]]
-    if data_morphed[0] is None:
-        if data_morphed[1] is None:
-            data = np.r_[[], []]
-            vertices = [np.array([], int), np.array([], int)]
-        else:
-            data = data_morphed[1]
-            vertices = [np.array([], int), vertices[1]]
-    elif data_morphed[1] is None:
-        data = data_morphed[0]
-        vertices = [vertices[0], np.array([], int)]
-    else:
-        data = np.r_[data_morphed[0], data_morphed[1]]
-
-    if isinstance(stc_from, VectorSourceEstimate):
-        stc_to = VectorSourceEstimate(data, vertices, stc_from.tmin,
-                                      stc_from.tstep, subject=subject_to,
-                                      verbose=stc_from.verbose)
-    else:
-        stc_to = SourceEstimate(data, vertices, stc_from.tmin, stc_from.tstep,
-                                subject=subject_to, verbose=stc_from.verbose)
-    logger.info('[done]')
+    source_morph = SourceMorph(None, subject_from=subject_from,
+                               subject_to=subject_to, spacing=grade,
+                               smooth=smooth, subjects_dir=subjects_dir,
+                               warn=warn, verbose=verbose)
+    source_morph.kind = 'surface'
+
+    # SourceSpace info
+    hemis = [0, 1]
+    for hemi in hemis:
+        source_morph.data.update({str(hemi): stc_from.vertices[hemi]})
+    source_morph.data.update({'hemis': hemis})
+    source_morph.data.update(_compute_morph_data(source_morph,
+                                                 verbose=verbose))
```

Comment:
this is a code smell. You create an object but then you need to update its internal state without using a method of the object. Can we avoid this?